File: typescript/packages/protocol/src/terminal.ts
import { ProtocolError } from "./errors.js";

export const TERMINAL_EVENT_VERSION = 0x01;
export const MAX_TERMINAL_COLUMNS = 1_000;
export const MAX_TERMINAL_ROWS = 500;
export const MAX_TERMINAL_CELLS = 131_072;
export const TERMINAL_EVENT_HEADER_BYTES = 2;

const SIZE_EVENT_BYTES = TERMINAL_EVENT_HEADER_BYTES + 4;
const EXITED_EVENT_BYTES = TERMINAL_EVENT_HEADER_BYTES + 5;

const DATA = 0x01;
const RESIZE = 0x02;
const STARTED = 0x03;
const EXITED = 0x04;
const HEARTBEAT = 0x05;

export type TerminalInputEvent =
  | { readonly type: "data"; readonly data: Uint8Array }
  | { readonly type: "resize"; readonly columns: number; readonly rows: number };

export type TerminalOutputEvent =
  | { readonly type: "data"; readonly data: Uint8Array }
  | { readonly type: "resize"; readonly columns: number; readonly rows: number }
  | { readonly type: "started"; readonly columns: number; readonly rows: number }
  | {
    readonly type: "exited";
    readonly status: number;
    readonly outputTruncated: boolean;
  }
  | { readonly type: "heartbeat" };

export function encodeTerminalInputEvent(event: TerminalInputEvent): Uint8Array {
  switch (event.type) {
    case "data":
      return encodeData(DATA, event.data);
    case "resize":
      return encodeSize(RESIZE, event.columns, event.rows);
  }
}

export function decodeTerminalInputEvent(payload: Uint8Array): TerminalInputEvent {
  const type = terminalEventType(payload);
  switch (type) {
    case DATA:
      return { type: "data", data: payload.subarray(TERMINAL_EVENT_HEADER_BYTES) };
    case RESIZE:
      return { type: "resize", ...decodeSize(payload, "resize") };
    default:
      throw unknownType("input", type);
  }
}

export function encodeTerminalOutputEvent(event: TerminalOutputEvent): Uint8Array {
  switch (event.type) {
    case "data":
      return encodeData(DATA, event.data);
    case "resize":
      return encodeSize(RESIZE, event.columns, event.rows);
    case "started":
      return encodeSize(STARTED, event.columns, event.rows);
    case "exited": {
      requireInt32(event.status, "terminal exit status");
      const payload = eventHeader(EXITED, EXITED_EVENT_BYTES);
      new DataView(payload.buffer).setInt32(TERMINAL_EVENT_HEADER_BYTES, event.status);
      payload[TERMINAL_EVENT_HEADER_BYTES + 4] = event.outputTruncated ? 1 : 0;
      return payload;
    }
    case "heartbeat":
      return eventHeader(HEARTBEAT, TERMINAL_EVENT_HEADER_BYTES);
  }
}

export function decodeTerminalOutputEvent(payload: Uint8Array): TerminalOutputEvent {
  const type = terminalEventType(payload);
  switch (type) {
    case DATA:
      return { type: "data", data: payload.subarray(TERMINAL_EVENT_HEADER_BYTES) };
    case RESIZE:
      return { type: "resize", ...decodeSize(payload, "resize") };
    case STARTED:
      return { type: "started", ...decodeSize(payload, "started") };
    case EXITED:
      requireLength(payload, EXITED_EVENT_BYTES, "exited");
      if (payload[TERMINAL_EVENT_HEADER_BYTES + 4]! > 1) {
        throw new ProtocolError(
          "invalid_terminal_exit_flags",
          `invalid terminal exited flags 0x${hex(payload[TERMINAL_EVENT_HEADER_BYTES + 4])}`,
        );
      }
      return {
        type: "exited",
        status: new DataView(
          payload.buffer,
          payload.byteOffset,
          payload.byteLength,
        ).getInt32(TERMINAL_EVENT_HEADER_BYTES),
        outputTruncated: payload[TERMINAL_EVENT_HEADER_BYTES + 4] === 1,
      };
    case HEARTBEAT:
      requireLength(payload, TERMINAL_EVENT_HEADER_BYTES, "heartbeat");
      return { type: "heartbeat" };
    default:
      throw unknownType("output", type);
  }
}

function encodeData(type: number, data: Uint8Array): Uint8Array {
  const payload = eventHeader(type, TERMINAL_EVENT_HEADER_BYTES + data.byteLength);
  payload.set(data, TERMINAL_EVENT_HEADER_BYTES);
  return payload;
}

function encodeSize(
  type: number,
  columns: number,
  rows: number,
): Uint8Array {
  validateTerminalSize(columns, rows);
  const payload = eventHeader(type, SIZE_EVENT_BYTES);
  const view = new DataView(payload.buffer);
  view.setUint16(TERMINAL_EVENT_HEADER_BYTES, columns);
  view.setUint16(TERMINAL_EVENT_HEADER_BYTES + 2, rows);
  return payload;
}

function decodeSize(
  payload: Uint8Array,
  name: string,
): { readonly columns: number; readonly rows: number } {
  requireLength(payload, SIZE_EVENT_BYTES, name);
  const view = new DataView(
    payload.buffer,
    payload.byteOffset,
    payload.byteLength,
  );
  const columns = view.getUint16(TERMINAL_EVENT_HEADER_BYTES);
  const rows = view.getUint16(TERMINAL_EVENT_HEADER_BYTES + 2);
  validateTerminalSize(columns, rows);
  return { columns, rows };
}

function eventHeader(type: number, length: number): Uint8Array {
  const payload = new Uint8Array(length);
  payload[0] = TERMINAL_EVENT_VERSION;
  payload[1] = type;
  return payload;
}

function terminalEventType(payload: Uint8Array): number {
  if (payload.byteLength < TERMINAL_EVENT_HEADER_BYTES) {
    throw new ProtocolError(
      "terminal_event_truncated",
      "terminal event must contain a version and type",
    );
  }
  const version = payload[0];
  if (version !== TERMINAL_EVENT_VERSION) {
    throw new ProtocolError(
      "unknown_terminal_event_version",
      `unknown terminal event version 0x${hex(version)}`,
    );
  }
  return payload[1]!;
}

function requireLength(payload: Uint8Array, expected: number, name: string): void {
  if (payload.byteLength !== expected) {
    throw new ProtocolError(
      "invalid_terminal_event_length",
      `${name} terminal event is ${payload.byteLength} bytes; expected ${expected}`,
    );
  }
}

export function validateTerminalSize(columns: number, rows: number): void {
  requireDimension(columns, "terminal columns", MAX_TERMINAL_COLUMNS);
  requireDimension(rows, "terminal rows", MAX_TERMINAL_ROWS);
  if (columns * rows > MAX_TERMINAL_CELLS) {
    throw new ProtocolError(
      "invalid_terminal_dimension",
      `terminal viewport must not exceed ${MAX_TERMINAL_CELLS} cells`,
    );
  }
}

function requireDimension(value: number, name: string, maximum: number): void {
  if (!Number.isInteger(value) || value < 1 || value > maximum) {
    throw new ProtocolError(
      "invalid_terminal_dimension",
      `${name} must be an integer between 1 and ${maximum}`,
    );
  }
}

function requireInt32(value: number, name: string): void {
  if (!Number.isInteger(value) || value < -0x8000_0000 || value > 0x7fff_ffff) {
    throw new ProtocolError(
      "invalid_terminal_exit_status",
      `${name} must fit in a signed 32-bit integer`,
    );
  }
}

function unknownType(direction: string, type: number): ProtocolError {
  return new ProtocolError(
    "unknown_terminal_event_type",
    `unknown terminal ${direction} event type 0x${hex(type)}`,
  );
}

function hex(value: number | undefined): string {
  return (value ?? 0).toString(16).padStart(2, "0");
}

File: rust/src/protocol/terminal.rs
//! Versioned terminal input and output events carried by byte records.

use thiserror::Error;

/// Current terminal event payload version.
pub const TERMINAL_EVENT_VERSION: u8 = 0x01;
/// Largest supported terminal width.
pub const MAX_TERMINAL_COLUMNS: u16 = 1_000;
/// Largest supported terminal height.
pub const MAX_TERMINAL_ROWS: u16 = 500;
/// Largest supported terminal viewport area.
pub const MAX_TERMINAL_CELLS: u32 = 131_072;

const HEADER_LEN: usize = 2;
const FIXED_EVENT_LEN: usize = HEADER_LEN + 4;
const DATA: u8 = 0x01;
const RESIZE: u8 = 0x02;
const STARTED: u8 = 0x03;
const EXITED: u8 = 0x04;
const HEARTBEAT: u8 = 0x05;

/// One event sent from a terminal controller to its PTY host.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TerminalInputEvent<'a> {
    /// Bytes to write to the PTY.
    Data(&'a [u8]),
    /// Requested PTY dimensions.
    Resize {
        /// Terminal columns.
        columns: u16,
        /// Terminal rows.
        rows: u16,
    },
}

/// One event sent from a PTY host to terminal observers.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TerminalOutputEvent<'a> {
    /// Bytes emitted by the PTY.
    Data(&'a [u8]),
    /// PTY dimensions accepted from a controller.
    Resize {
        /// Terminal columns.
        columns: u16,
        /// Terminal rows.
        rows: u16,
    },
    /// The PTY child started with these dimensions.
    Started {
        /// Terminal columns.
        columns: u16,
        /// Terminal rows.
        rows: u16,
    },
    /// The PTY child exited with this status.
    Exited {
        /// Signed process status selected by the host.
        status: i32,
    },
    /// The PTY host is still connected.
    Heartbeat,
}

/// Encodes one terminal input event.
pub fn encode_terminal_input(
    event: TerminalInputEvent<'_>,
) -> Result<Vec<u8>, TerminalProtocolError> {
    match event {
        TerminalInputEvent::Data(data) => Ok(encode_data(DATA, data)),
        TerminalInputEvent::Resize { columns, rows } => encode_size(RESIZE, columns, rows),
    }
}

/// Decodes one terminal input event.
pub fn decode_terminal_input(
    payload: &[u8],
) -> Result<TerminalInputEvent<'_>, TerminalProtocolError> {
    match event_type(payload)? {
        DATA => Ok(TerminalInputEvent::Data(&payload[HEADER_LEN..])),
        RESIZE => {
            let (columns, rows) = decode_size(payload, "resize")?;
            Ok(TerminalInputEvent::Resize { columns, rows })
        }
        event_type => Err(TerminalProtocolError::UnknownType {
            direction: "input",
            event_type,
        }),
    }
}

/// Encodes one terminal output event.
pub fn encode_terminal_output(
    event: TerminalOutputEvent<'_>,
) -> Result<Vec<u8>, TerminalProtocolError> {
    match event {
        TerminalOutputEvent::Data(data) => Ok(encode_data(DATA, data)),
        TerminalOutputEvent::Resize { columns, rows } => encode_size(RESIZE, columns, rows),
        TerminalOutputEvent::Started { columns, rows } => encode_size(STARTED, columns, rows),
        TerminalOutputEvent::Exited { status } => {
            let mut payload = event_header(EXITED, FIXED_EVENT_LEN);
            payload[HEADER_LEN..].copy_from_slice(&status.to_be_bytes());
            Ok(payload)
        }
        TerminalOutputEvent::Heartbeat => Ok(event_header(HEARTBEAT, HEADER_LEN)),
    }
}

/// Decodes one terminal output event.
pub fn decode_terminal_output(
    payload: &[u8],
) -> Result<TerminalOutputEvent<'_>, TerminalProtocolError> {
    match event_type(payload)? {
        DATA => Ok(TerminalOutputEvent::Data(&payload[HEADER_LEN..])),
        RESIZE => {
            let (columns, rows) = decode_size(payload, "resize")?;
            Ok(TerminalOutputEvent::Resize { columns, rows })
        }
        STARTED => {
            let (columns, rows) = decode_size(payload, "started")?;
            Ok(TerminalOutputEvent::Started { columns, rows })
        }
        EXITED => {
            require_length(payload, FIXED_EVENT_LEN, "exited")?;
            let status = i32::from_be_bytes([
                payload[HEADER_LEN],
                payload[HEADER_LEN + 1],
                payload[HEADER_LEN + 2],
                payload[HEADER_LEN + 3],
            ]);
            Ok(TerminalOutputEvent::Exited { status })
        }
        HEARTBEAT => {
            require_length(payload, HEADER_LEN, "heartbeat")?;
            Ok(TerminalOutputEvent::Heartbeat)
        }
        event_type => Err(TerminalProtocolError::UnknownType {
            direction: "output",
            event_type,
        }),
    }
}

fn encode_data(event_type: u8, data: &[u8]) -> Vec<u8> {
    let mut payload = event_header(event_type, HEADER_LEN + data.len());
    payload[HEADER_LEN..].copy_from_slice(data);
    payload
}

fn encode_size(event_type: u8, columns: u16, rows: u16) -> Result<Vec<u8>, TerminalProtocolError> {
    validate_terminal_size(columns, rows)?;
    let mut payload = event_header(event_type, FIXED_EVENT_LEN);
    payload[HEADER_LEN..HEADER_LEN + 2].copy_from_slice(&columns.to_be_bytes());
    payload[HEADER_LEN + 2..].copy_from_slice(&rows.to_be_bytes());
    Ok(payload)
}

fn decode_size(payload: &[u8], name: &'static str) -> Result<(u16, u16), TerminalProtocolError> {
    require_length(payload, FIXED_EVENT_LEN, name)?;
    let columns = u16::from_be_bytes([payload[HEADER_LEN], payload[HEADER_LEN + 1]]);
    let rows = u16::from_be_bytes([payload[HEADER_LEN + 2], payload[HEADER_LEN + 3]]);
    validate_terminal_size(columns, rows)?;
    Ok((columns, rows))
}

fn event_header(event_type: u8, len: usize) -> Vec<u8> {
    let mut payload = vec![0; len];
    payload[0] = TERMINAL_EVENT_VERSION;
    payload[1] = event_type;
    payload
}

fn event_type(payload: &[u8]) -> Result<u8, TerminalProtocolError> {
    if payload.len() < HEADER_LEN {
        return Err(TerminalProtocolError::Truncated);
    }
    if payload[0] != TERMINAL_EVENT_VERSION {
        return Err(TerminalProtocolError::UnknownVersion(payload[0]));
    }
    Ok(payload[1])
}

fn require_length(
    payload: &[u8],
    expected: usize,
    name: &'static str,
) -> Result<(), TerminalProtocolError> {
    if payload.len() != expected {
        return Err(TerminalProtocolError::InvalidLength {
            name,
            actual: payload.len(),
            expected,
        });
    }
    Ok(())
}

/// Validates terminal dimensions accepted by every TSF implementation.
pub fn validate_terminal_size(columns: u16, rows: u16) -> Result<(), TerminalProtocolError> {
    if columns == 0 || columns > MAX_TERMINAL_COLUMNS {
        return Err(TerminalProtocolError::InvalidDimension {
            name: "columns",
            maximum: MAX_TERMINAL_COLUMNS,
        });
    }
    if rows == 0 || rows > MAX_TERMINAL_ROWS {
        return Err(TerminalProtocolError::InvalidDimension {
            name: "rows",
            maximum: MAX_TERMINAL_ROWS,
        });
    }
    if u32::from(columns) * u32::from(rows) > MAX_TERMINAL_CELLS {
        return Err(TerminalProtocolError::TooManyCells {
            maximum: MAX_TERMINAL_CELLS,
        });
    }
    Ok(())
}

/// A terminal event payload is malformed or unsupported.
#[derive(Clone, Debug, Eq, Error, PartialEq)]
pub enum TerminalProtocolError {
    /// The payload does not contain its version and type.
    #[error("terminal event must contain a version and type")]
    Truncated,
    /// The payload uses an unknown version.
    #[error("unknown terminal event version 0x{0:02x}")]
    UnknownVersion(u8),
    /// The payload uses an unknown direction-specific event type.
    #[error("unknown terminal {direction} event type 0x{event_type:02x}")]
    UnknownType {
        /// Event direction.
        direction: &'static str,
        /// Unknown type byte.
        event_type: u8,
    },
    /// A fixed-width event has the wrong length.
    #[error("{name} terminal event is {actual} bytes; expected {expected}")]
    InvalidLength {
        /// Event name.
        name: &'static str,
        /// Actual encoded length.
        actual: usize,
        /// Required encoded length.
        expected: usize,
    },
    /// A terminal dimension is outside its supported range.
    #[error("terminal {name} must be between 1 and {maximum}")]
    InvalidDimension {
        /// Dimension name.
        name: &'static str,
        /// Inclusive maximum.
        maximum: u16,
    },
    /// A terminal viewport has too many cells.
    #[error("terminal viewport must not exceed {maximum} cells")]
    TooManyCells {
        /// Inclusive maximum.
        maximum: u32,
    },
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use super::*;

    #[derive(Deserialize)]
    struct TestVector {
        name: String,
        bytes: Vec<u8>,
    }

    #[test]
    fn matches_shared_test_vectors() {
        let vectors: Vec<TestVector> =
            serde_json::from_str(include_str!("../../testdata/terminal-events.json"))
                .expect("terminal event vectors");
        for vector in vectors {
            let encoded = match vector.name.as_str() {
                "input-data" => encode_terminal_input(TerminalInputEvent::Data(&[0, 1, 255])),
                "input-resize" => encode_terminal_input(TerminalInputEvent::Resize {
                    columns: 132,
                    rows: 43,
                }),
                "output-data" => encode_terminal_output(TerminalOutputEvent::Data(&[27, 91, 109])),
                "output-resize" => encode_terminal_output(TerminalOutputEvent::Resize {
                    columns: 80,
                    rows: 24,
                }),
                "output-started" => encode_terminal_output(TerminalOutputEvent::Started {
                    columns: 120,
                    rows: 40,
                }),
                "output-exited" => {
                    encode_terminal_output(TerminalOutputEvent::Exited { status: -1 })
                }
                "output-heartbeat" => encode_terminal_output(TerminalOutputEvent::Heartbeat),
                name => panic!("unknown terminal test vector {name}"),
            }
            .expect("encode terminal event");
            assert_eq!(encoded, vector.bytes, "{}", vector.name);
        }
    }

    #[test]
    fn round_trips_input_events() {
        for event in [
            TerminalInputEvent::Data(&[0, 1, 255]),
            TerminalInputEvent::Resize {
                columns: 132,
                rows: 43,
            },
        ] {
            let encoded = encode_terminal_input(event).expect("encode input event");
            assert_eq!(decode_terminal_input(&encoded), Ok(event));
        }
    }

    #[test]
    fn round_trips_output_events() {
        for event in [
            TerminalOutputEvent::Data(&[27, 91, 109]),
            TerminalOutputEvent::Resize {
                columns: 80,
                rows: 24,
            },
            TerminalOutputEvent::Started {
                columns: 120,
                rows: 40,
            },
            TerminalOutputEvent::Exited { status: -1 },
            TerminalOutputEvent::Heartbeat,
        ] {
            let encoded = encode_terminal_output(event).expect("encode output event");
            assert_eq!(decode_terminal_output(&encoded), Ok(event));
        }
    }

    #[test]
    fn rejects_invalid_events() {
        assert_eq!(
            decode_terminal_input(&[TERMINAL_EVENT_VERSION]),
            Err(TerminalProtocolError::Truncated)
        );
        assert_eq!(
            decode_terminal_output(&[2, DATA]),
            Err(TerminalProtocolError::UnknownVersion(2))
        );
        assert!(matches!(
            decode_terminal_input(&[TERMINAL_EVENT_VERSION, RESIZE, 0, 80, 0]),
            Err(TerminalProtocolError::InvalidLength { .. })
        ));
        assert_eq!(
            encode_terminal_input(TerminalInputEvent::Resize {
                columns: 0,
                rows: 24
            }),
            Err(TerminalProtocolError::InvalidDimension {
                name: "columns",
                maximum: MAX_TERMINAL_COLUMNS,
            })
        );
        assert_eq!(
            encode_terminal_output(TerminalOutputEvent::Started {
                columns: MAX_TERMINAL_COLUMNS,
                rows: MAX_TERMINAL_ROWS,
            }),
            Err(TerminalProtocolError::TooManyCells {
                maximum: MAX_TERMINAL_CELLS,
            })
        );
    }
}
